# ORFcalling
This repository provides code to reproduce parts of results from RiboCode paper

Comparison with RiboTaper and ORFscore on exon level
----------------

This section is for reproducing Fig2.C results. 

1. run RiboTaper to create the ribosome profiling track data (true positive data) and mRNA track data (true negative data)

   create annotation data

   `create_annotations_files.bash gtf_file fasta_file true fasle anotation_dir` 

   `Ribotaper.bash rpf.bam mRNA.bam annotation_dir 27,28 12,12 8`

   Three files will be used in following steps:

   *P_sites_all_tracks_ccds* :

   *Centered_RNA_tracks_ccds*:

   results_ccds

   *annot/frames_ccds*:

   we provide these file  in the "data" folder. 

2. calculating the p values using RiboCode:

   Note: change the 

   `python generate_ribocode_pvalue.py`

3. extract the P-values generated by RiboTaper algorithm.

   `python generate_ROC_input.py`

4. generate ROC curves

   `Rscript ROC.R`



Comparison with Rp-Bp, ORF-RATER, RiboORF
--------
1. STAR map results for RPF and mRNA

2. We randomly select 1000 protein coding genes as the test dataset 

   `python generate_simulatedFq.py`

3. change the "protein_coding" as "non-conding" in GTF as the 

   `python generate_simulatedGTF.py`

4. mapping using STAR

   build index

   `STAR --runThreadN 8 --runMode genomeGenerate --genomeFastaFiles $genome_fa_file --sjdbGTFfile simulated.gtf --genomeDir staridx`

   mapping

   `STAR --outFilterType BySJout --runThreadN 8 --outFilterMismatchNmax 2 --genomeDir staridx --readFilesIn simulated.fq --outFileNamePrefix simulated --outSAMtype BAM SortedByCoordinate --quantMode TranscriptomeSAM GeneCounts --outFilterMultimapNmax 1 --outFilterMatchNmin 16 --alignEndsType EndToEnd --outSAMattributes All`

5. Run different methods:

   Rp-Bp:

   `prepare-rpbp-genome prepare.yaml --num-cpus 7 --mem 100G --overwrite --logging-level INFO`

   `run-all-rpbp-instances config.yaml --overwrite --num-cpus 16 --logging-level INFO --log-file log.alignments-only.txt --keep-intermediate-files`

   ORFrate:

   see the bash file in ORFrate folder

   RiboORF:

   see the bash file in RiboORF folder

   RiboCode:

   `RiboCode -a annot -c config.txt -l yes -p 1 -m 9 -o ribocode_result`

6. generate roc data:

   for each method, we can generate the ROC using the "" script in each folder.

   ​

   ​

   ​

   ​

   ​

   ​

   ​

