# ORFcalling
This repository provides code to reproduce parts of results from RiboCode paper, all the data used in this instruction is stored in Google Drive: https://drive.google.com/open?id=1Ibpeu9r-TpCce4qQG-gwZvj4XRpR7RKn

Comparison with RiboTaper and ORFscore on exon level
----------------

This section is for reproducing Fig2.C results.

1. run RiboTaper to create the ribosome profiling track data (true positive data) and mRNA track data (true negative data)

   create annotation data

   `create_annotations_files.bash genecode.v19.annotation.gtf hg19_genome.fa true true ribotaper_annot` 

   `Ribotaper.bash mapping_files/rpf_aligned.bam mapping_files/mRNA_aligned.bam ribotaper_annot 26,27,28,29 12,12,12,12 8`

   Three files generated by *Ribotaper* will be used in following steps:

   - P_sites_all_tracks_ccds


   - Centered_RNA_tracks_ccds
   - results_ccds
   - annot/frames_ccds

   These files can be downloaded from Google driver:

2. calculating the p values using RiboCode:

   Note: please change the path of input files in script

   `python generate_ribocode_pvalue.py`

   This step will generate "ribocode_result.txt".

3. extract the P-values generated by RiboTaper algorithm and generate the ROC input file

   `python generate_ROC_input.py`

   This step generates "ROC_input.txt"

4. generate ROC curves

   `Rscript ROC.R`

   This step will generate "roc.pdf" and "auc.txt". 


Comparison with Rp-Bp, ORF-RATER, RiboORF
--------

This section is for reproducing Fig3.a results

1. Preparing the STAR mapping results for RPF and mRNA：

   Please see in the folder: mapping_files on google drive.

2. We randomly select 1000 protein coding genes as the test dataset 

   `python generate_simulatedFq.py`

   This step will generate "selected.gtf" file and a simulated fastq file in which the rpf reads of the selected genes will be replaced by the mRNA reads.

3. change the type of selected genes fro "protein_coding" to "lincRNA" in GTF

   `python generate_simulatedGTF.py`

4. mapping using STAR

   `./run_star.sh`

5. Run different methods:

   - Rp-Bp: 
   - ORFrate:
   - RiboORF:
   - RiboCode:

6. Generate the ROC data:

   `python generate_ROCdata.py`

   `./roc.sh`

   ​

   ​

   ​

   ​

   ​

   ​

   ​

   ​

